### PR TITLE
Partially styled elements from a paragraph are not correctly shown in XWiki PDF export #113

### PR DESCRIPTION
**Pull request by oanalavinia**
* if a paragraph has partially styled nodes, fallback to display only the text and ignore any bold, links.. elements
* refactoring is added in a separate commit

### CODE REVIEW

File: application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml [hunk 1]
```diff
@@ -494,13 +502,38 @@ define('diagram-store', ['jquery', 'xwiki-meta', 'xwiki-utils', 'diagram-utils',
...
+    return ['ul', 'ol'].includes(tag);
+  }
+
+  var isInlineElement = function(element) {
+    var tag = element.tagName.toLowerCase();
+    return ['b', 'i', 'span'].includes(tag);
+  }
+
+  var isParagraphWithPartialStyle = function(node, nbOfSiblings) {
+    if (node.nodeType === 3 || (isInlineElement(node) &amp;&amp; nbOfSiblings &gt; 1)) {
```
Comment (mflorea) [inline application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml]: You can return directly the condition, no need for an ``IF``.

File: application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml [hunk 2]
```diff
@@ -642,24 +674,35 @@ define('diagram-link-editor', [
...
         }
 
-        var previousHeight;
         if (this.foAltText != null) {
           var s = this.state;
           var htmlConverter = document.createElement('textarea');
           var content = document.createElementNS('http://www.w3.org/2000/svg', 'g');
-          var elements = fo.firstElementChild.firstElementChild.childElements();
+          var elementsContainer = $(fo).children().first().children().first();
+          var elements = elementsContainer.contents();
```
Comment (mflorea) [inline application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml]: From the documentation of ``contents()``:

> Get the children of each element in the set of matched elements, **including text and comment nodes**.

So if your variable is called ``elements`` then I assume you want only the elements and thus you should use ``children()`` instead. If you need to process all nodes, including text and comment nodes, then you should rename the variable.

Comment (oanalavinia) [reply application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml]: I do need to process all nodes, so I'll do the rename

File: application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml [hunk 3]
```diff
@@ -650,24 +679,35 @@ define('diagram-link-editor', [
...
+                if (svgHandler.isParagraphWithPartialStyle(node, childNodes.length)) {
+                  node = $('&lt;span&gt;&lt;/span&gt;').html(this.parentElement.innerText)[0];
+                  stop = true;
                 }
-                previousHeight = h;
-                var alt = svgHandler.convertTextElementToSvg(element, s, w, h, htmlConverter, str, rectWidth);
+                var alt = svgHandler.convertTextElementToSvg(node, s, w, h, htmlConverter, str, rectWidth, index == 0);
                 content.append(alt);
+                // If we took the inner text of the parent, the other child nodes are not needed.
+                if (stop) {
```
Comment (mflorea) [inline application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml]: You can return directly ``!stop``.

**Pull request merged**